pkg/k8s/finalizer.go

```diff
@@ -0,0 +1,84 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package k8s
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type FinalizerManager interface {
+	AddFinalizers(ctx context.Context, object client.Object, finalizers ...string) error
+	RemoveFinalizers(ctx context.Context, object client.Object, finalizers ...string) error
+}
+
+func NewDefaultFinalizerManager(k8sClient client.Client, log logr.Logger) FinalizerManager {
+	return &defaultFinalizerManager{
+		k8sClient: k8sClient,
+		log:       log,
+	}
+}
+
+type defaultFinalizerManager struct {
+	k8sClient client.Client
+	log       logr.Logger
+}
+
+func (m *defaultFinalizerManager) AddFinalizers(ctx context.Context, obj client.Object, finalizers ...string) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := m.k8sClient.Get(ctx, NamespacedName(obj), obj); err != nil {
+			return err
+		}
+
+		oldObj := obj.DeepCopyObject().(client.Object)
+		needsUpdate := false
+		for _, finalizer := range finalizers {
+			if !controllerutil.ContainsFinalizer(obj, finalizer) {
+				m.log.Info("adding finalizer", "object", obj.GetObjectKind().GroupVersionKind().Kind, "name", obj.GetName(), "finalizer", finalizer)
+				controllerutil.AddFinalizer(obj, finalizer)
+				needsUpdate = true
+			}
+		}
+		if !needsUpdate {
+			return nil
+		}
+		return m.k8sClient.Patch(ctx, obj, client.MergeFromWithOptions(oldObj, client.MergeFromWithOptimisticLock{}))
+	})
+}
+
+func (m *defaultFinalizerManager) RemoveFinalizers(ctx context.Context, obj client.Object, finalizers ...string) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := m.k8sClient.Get(ctx, NamespacedName(obj), obj); err != nil {
+			return err
+		}
+
+		oldObj := obj.DeepCopyObject().(client.Object)
+		needsUpdate := false
+		for _, finalizer := range finalizers {
```

Comment:
I am just boldly thinking if we should remove all finalizers although we should have only one so far, in stead of checking every one. We get into here because we are cleaning up resources and removing the blocker of deletion. I don't see any reason we can tolerate finalizer(s) leftover after successfully making this call. WDYT?

Reply:
Do you mean we need not check against the `finalizers` passed to the function and instead remove all finalizers here?

I don't think we should do that within the generic finalizer manager. If we want to remove all finalizers in the CNINode reconciler, then we should pass in all finalizers to the function.

Reply:
Resolving the thread to merge the PR, I'll follow up on this offline.